APCAppCore/APCAppCore/Library/Scoring/APCScoring.m

```diff
@@ -575,113 +735,148 @@ - (NSArray *)retrieveResultSummaryFromResults:(NSSet *)results latestOnly:(BOOL)
     return allResultSummaries;
 }
 
-- (void)groupDatasetByDay
+- (void)groupByPeriod:(APHTimelineGroups)period
 {
-    NSMutableArray *groupedDataset = [NSMutableArray array];
-    NSArray *days = [self.dataPoints valueForKeyPath:@"@distinctUnionOfObjects.datasetDateKey"];
+    NSString *periodKeyPath = nil;
+    
+    switch (period) {
+        case APHTimelineGroupWeek:
+            periodKeyPath = kDatasetGroupByWeek;
+            break;
+        case APHTimelineGroupMonth:
+            periodKeyPath = kDatasetGroupByMonth;
+            break;
+        case APHTimelineGroupYear:
+            periodKeyPath = kDatasetGroupByYear;
+            break;
+        default:
+            periodKeyPath = kDatasetGroupByDay;
+            break;
+    }
+    
+    // Group the dataset based on the provided period.
+    NSDictionary *groupedDataset = [self groupByKeyPath:periodKeyPath dataset:self.dataPoints];
     
-    for (NSString *day in days) {
-        NSMutableDictionary *entry = [NSMutableDictionary dictionary];
-        [entry setObject:day forKey:kDatasetDateKey];
+    // Summarize data for each group.
+    NSDictionary *summarizedDataset = [self summarizeDataset:groupedDataset period:period];
+    NSArray *sortedSummaryKeys = [summarizedDataset.allKeys sortedArrayUsingSelector:@selector(compare:)];
```

Comment:
The method -summarizeDataset:period: return an NSDictionary. That dictionary uses NSDate as the key and holds NSDictionary as the value. As such the structure is unordered. We need the keys in order by the date, thus we are sorting the keys in returned dictionary (summarizedDataset).